File: backend/src/models/ApiKey.ts
import mongoose, { Schema, Document } from 'mongoose';
import { IUser } from './User';

export interface IApiKey extends Document {
  key: string;
  secret: string;
  name: string;
  ownerId: IUser['_id'];
  lastUsedAt?: Date;
  expiresAt?: Date;
  isActive: boolean;
  createdAt: Date;
  updatedAt: Date;
}

const ApiKeySchema: Schema = new Schema({
  key: { type: String, required: true, unique: true },
  secret: { type: String, required: true },
  name: { type: String, required: true, trim: true },
  ownerId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  lastUsedAt: { type: Date },
  expiresAt: { type: Date },
  isActive: { type: Boolean, default: true },
}, { timestamps: true });

ApiKeySchema.index({ ownerId: 1 });
ApiKeySchema.index({ key: 1 });

export default mongoose.model<IApiKey>('ApiKey', ApiKeySchema);

File: backend/app/models/api_key.py
from datetime import datetime, timezone
from typing import Optional
from beanie import Document, PydanticObjectId
from pydantic import Field


class ApiKey(Document):
    key: str = Field(unique=True)
    secret: str
    name: str
    owner_id: PydanticObjectId = Field(alias="ownerId")
    last_used_at: Optional[datetime] = Field(default=None, alias="lastUsedAt")
    expires_at: Optional[datetime] = Field(default=None, alias="expiresAt")
    is_active: bool = Field(default=True, alias="isActive")
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "apikeys"
        populate_by_name = True
        indexes = [
            "owner_id",
            "key",
        ]
